
**July 22nd - 28th 2020**

## **Note:** You are expected to be familiar with Rstudio, RMarkdown documents and Bioconductor.
## If you are not then you are advised to work through this material **before** Summer School
- [Introduction to Bioconductor](Introduction/bioc-intro.html).  
- [Bioconductor answers](Introduction/bioc-intro.solutions.html).  
  + [source](Introduction/bioc-intro.Rmd)
  + [Source with solutions](Introduction/bioc-intro.solutions.Rmd). 

# Day 1

**Please note 9.00 Start**
- 09:00 - 09:40; Welcome & Introduction to Functional Genomics (Mark & Rory)
- 09:40 - 12:30; Data Processing for Next Generation Sequencing (Joanna & Shoko)
    + Introduction to next generation sequencing 
    + A brief Introduction to genomic file formats 
    + Quality control and artefact removal 
    + Practical - Learn to use FastQC and Cutadapt on a sample dataset
    + Short Read alignment and quality control 
    + Practical - Alignment to reference genomes using BWA and STAR
    
- 12:30 - 13:30; LUNCH BREAK

- 13:30 - 17:00; Bulk RNAseq (Ash, Abbi & Chandu)
    + Introduction to RNAseq Analysis in R
    + RNA-seq Pre-processing 
    
# Day 2

- 09:30 - 17:00; Bulk RNAseq Continued (Ash, Dom & Chandu)
    + Statistical Analysis of Bulk RNAseq Data (Dom)
    + Experimental Design of Bulk RNAseq studies (Chandu)
    + Differential Expression for RNA-seq  (Ash)
    
- 12:30 - 13:30; LUNCH

# Day 3

- 09:30 - 12:30; Bulk RNAseq (cont.) (Ash & Abi)
    + Annotation and Visualisation of RNA-seq results (Abbi)
    + Gene-set testing (Ash)

- 12:30 - 13:30; LUNCH BREAK

- 13:30 - 17:00; Single Cell RNAseq (Stephane, & Kasia)
    + scRNA-seq - introduction, inc design
    + scRNA-seq - preprocessing, inc cell calling
    + scRNA-seq - normalisation, visualisation, batch correction
    
**SOCIAL: Zoom Talk: “Perspectives in AI for Cancer Bioinformatics by Rory Stark”. **)

# ---- Weekend ----

# Day 4

- 09:30 - 17:00; Single Cell RNAseq (cont.) (Stephane & Zeynep)
    + scRNA-seq - feature selection and dimension reduction
    + scRNA-seq - clustering and marker gene detection 
    + scRNA-seq - differential expression and differential abundance 
    
- 12:30 - 13:30; LUNCH BREAK
  
# Day 5 

- 09:30 - 17:00; ChIPseq (Joanna & Shoko)
    + Introduction to ChIP-seq 
    + Introduction to Peak Calling 
    + Differential binding analysis 
    + Quality control methods for ChIP-seq 
    + Downstream analysis of ChIP-seq 
    + Introduction to (epi)Genomic data integration(
    + Chromatin accessibility and nucleosome occupancy with ATAC-seq 
    + Useful software utilities for the analysis of genomic data 

- 12:30 - 13:30; LUNCH BREAK

** End of CRUK CC Bioinformatics Summer School 2020**
-------------------------------------------------------------
